### PR TITLE
[Merged by Bors] - bootstrap will stop after 10 tries even if we didn't find anything, w…

### LINKED ISSUE #1997: Bootstrap takes forever when you know all peers.
## Description

we encountered this in testnet, after multiple discovery crawls we found a lot of peers and saved them in our `peer.json` though next time we go live, it is a prerequisite before we start connecting to neighbors we want to query our peers to find some new peers, we will keep querying and searching until we find at least 8 (configurable) new peers.  

the problem is, if we know all the network (due to it being quite small), we are going to keep searching forever and find the same peers all over again.

we need to decide on a termination policy even if we don't find new peers, this termination policy should take into consideration the opposite situation where there is a small number of peers and it takes a good amount of look ups to find new peers. (or a genesis where peers are just starting hence it takes time discovering them)

## Steps To Reproduce
- Start a network
- Start a peer and let it crawl the network over and over again and save the peer.json file.
- restart the node and see that you keep trying to discover new peers.

## Expected Behavior

stop searching after a few tries (need to find the sweet spot)

## Environment
Please complete the following information:
 - OS: 
 
 - Node Version: 
 
 - App Version (if applicable): 
 

## Additional Resources 

logs can be provided on demand.

### PR DESCRIPTION
**Pull request by y0sher**
## Motivation
We had an issue in testnet because the small number of peers, quite fast your addrbook is filled with all peers in the network and you can't find new nodes.
 currently a condition to finish bootstrap is that we found X new nodes, when you know all of them you won't find new nodes unless they join the network while you're doing this.

Closes #1997

## Changes
bootstrap will stop after 10 tries even if we didn't find anything, when the loaded addrbook is bigger than the required number of peers.

## Test Plan
added unit tests

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

### CODE REVIEW

File: p2p/discovery/refresher.go [hunk 1]
```diff
@@ -21,8 +21,12 @@ const (
 	lastQueriesCacheSize = 100
 
 	maxConcurrentRequests = 3
+
+	maxTries = 10
 )
 
+var defaultbackOffFunc = func(tries int) time.Duration { return time.Second * time.Duration(tries) }
```
Comment (noamnelke) [inline p2p/discovery/refresher.go]: Capitalization is weird, maybe `defaultBackoffFunc`?

File: p2p/discovery/refresher.go [hunk 2]
```diff
@@ -39,6 +43,8 @@ type refresher struct {
 	book      addressBook
 	bootNodes []*node.Info
 
+	backOffFunc func(tries int) time.Duration
```
Comment (noamnelke) [inline p2p/discovery/refresher.go]: Here too, I think "backoff" is a word, so the 'O' can be lowercase.

**Review by noamnelke — APPROVED** [review]
lgtm (only had a comment on capitalization)

**Comment by y0sher** [comment]
bors merge

**Comment by noamnelke** [comment]
bors merge

**Comment by noamnelke** [comment]
bors merge

**Pull request closed without merging**